### PR TITLE
Parametric Cross-Resonance

### PR DESCRIPTION
**Pull request by dowusu-antwi**
Augments `AceCR` to take a `theta` parameter (halved for echo implementation) (requires server-side changes for compilation):

image

### CODE REVIEW

**Comment by dowusu-antwi** [comment]
^ Response to review in https://github.com/SupertechLabs/cirq-superstaq/pull/421 updated here as well for parity.

File: qiskit_superstaq/custom_gates.py [hunk 1]
```diff
@@ -13,77 +13,79 @@ class AceCR(qiskit.circuit.Gate):
...
-        return AceCR(self.polarity, sandwich_rx_rads=-self.sandwich_rx_rads, label=self.label)
+        return AceCR(self.rads, -self.sandwich_rx_rads, label=self.label)
 
     def _define(self) -> None:
         qc = qiskit.QuantumCircuit(2, name=self.name)
-        first_sign = +1 if self.polarity == "+-" else -1
-        qc.rzx(first_sign * np.pi / 4, 0, 1)
+        if self.rads == 0:
+            self.definition = qc
+            return
```
Comment (richrines1) [inline qiskit_superstaq/custom_gates.py:56]: i don't think this is correct - the circuit should still have the `x(0)` and `rx(sandwich_rx_rads, 1)` gates. instead i would just do something like:
```python
if self.rads:
    qc.rzx(..)
qc.x(0)
if self.sandwich_rx_rads:
    qc.rx(..)
if self.rads:
    qx.rzx(..)
```

File: qiskit_superstaq/custom_gates.py [hunk 2]
```diff
@@ -13,77 +13,79 @@ class AceCR(qiskit.circuit.Gate):
...
-        )
+        return qiskit.quantum_info.Operator(self.definition).data
 
     def __repr__(self) -> str:
-        args = f"'{self.polarity}'"
-        if self.sandwich_rx_rads:
-            args += f", sandwich_rx_rads={self.sandwich_rx_rads}"
+        args = []
+        if self.rads not in [0, np.pi / 2]:
+            args.append(f"rads={self._rads_str}")
```
Comment (richrines1) [inline qiskit_superstaq/custom_gates.py:71]:
```suggestion
            args.append(f"rads={self.rads!r}")
```
(repr values shouldn't be rounded/include the unicode pi)

File: qiskit_superstaq/custom_gates.py [hunk 3]
```diff
@@ -13,77 +13,79 @@ class AceCR(qiskit.circuit.Gate):
...
+        if self.sandwich_rx_rads != 0:
+            args.append(f"sandwich_rx_rads={self.sandwich_rx_rads}")
```
Comment (richrines1) [inline qiskit_superstaq/custom_gates.py:73]:
```suggestion
            args.append(f"sandwich_rx_rads={self.sandwich_rx_rads!r}")
```

File: qiskit_superstaq/custom_gates.py [hunk 4]
```diff
@@ -13,77 +13,79 @@ class AceCR(qiskit.circuit.Gate):
...
         if self.label:
-            args += f", label='{self.label}'"
-        return f"qss.AceCR({args})"
+            args.append(f"label='{self.label}'")
```
Comment (richrines1) [inline qiskit_superstaq/custom_gates.py:75]:
```suggestion
            args.append(f"label={self.label!r}")
```
(the "!r" will add the quotes)

File: qiskit_superstaq/custom_gates.py [hunk 5]
```diff
@@ -13,77 +13,79 @@ class AceCR(qiskit.circuit.Gate):
...
+        return f"qss.AceCR({', '.join(args)})" if args else "qss.AceCR"
```
Comment (richrines1) [inline qiskit_superstaq/custom_gates.py:76]:
```suggestion
        return f"qss.AceCR({', '.join(args)})"
```
(repr should include parentheses)

File: qiskit_superstaq/custom_gates.py [hunk 6]
```diff
@@ -13,77 +13,79 @@ class AceCR(qiskit.circuit.Gate):
...
-            np.eye(2, dtype=dtype),
-        )
+        return qiskit.quantum_info.Operator(self.definition).data
 
     def __repr__(self) -> str:
-        args = f"'{self.polarity}'"
-        if self.sandwich_rx_rads:
-            args += f", sandwich_rx_rads={self.sandwich_rx_rads}"
+        args = []
+        if self.rads not in [0, np.pi / 2]:
```
Comment (richrines1) [inline qiskit_superstaq/custom_gates.py:70]: why is rads==0 treated the same way as rads==pi/2?

File: qiskit_superstaq/custom_gates.py [hunk 7]
```diff
@@ -13,77 +13,79 @@ class AceCR(qiskit.circuit.Gate):
...
-            super().__init__(name, 2, [], label=label)
-
-        self.polarity = polarity
+            name += "_rx"
+            params.append(sandwich_rx_rads)
+        super().__init__(name, 2, params, label=label)
+
+        rads_str = f"{rads}"
+        if np.isclose(round(rads / np.pi, 5) * np.pi, rads):
+            rads_str = f"{round(rads / np.pi, 5)}π"
```
Comment (richrines1) [inline qiskit_superstaq/custom_gates.py:44]: (assuming prior suggestions) imo this should be moved to `__str__` now that it's only used there

File: qiskit_superstaq/custom_gates.py [hunk 8]
```diff
@@ -13,77 +13,79 @@ class AceCR(qiskit.circuit.Gate):
...
-        return f"qss.AceCR({args})"
+            args.append(f"label='{self.label}'")
+        return f"qss.AceCR({', '.join(args)})" if args else "qss.AceCR"
 
     def __str__(self) -> str:
-        if not self.sandwich_rx_rads:
-            return f"AceCR{self.polarity}"
-        arg = qiskit.circuit.tools.pi_check(self.sandwich_rx_rads, ndigits=8, output="qasm")
-        return f"AceCR{self.polarity}|RXGate({arg})|"
+        if self.sandwich_rx_rads == 0 and self.rads in [0, np.pi / 2]:
```
Comment (richrines1) [inline qiskit_superstaq/custom_gates.py:79]: why is rads==0 treated the same way as rads==pi/2?

File: setup.cfg [hunk 9]
```diff
@@ -28,7 +28,6 @@ extend-ignore = E203, I100, I101, I201, I202  # these errors conflict with black
 [mypy]
 ignore_missing_imports = True
 show_error_codes = True
-disallow_any_generics = True
```
Comment (richrines1) [inline setup.cfg:31]: why this change? i think we want to keep this flag on

File: qiskit_superstaq/custom_gates.py (hunk 1 above)
Comment (dowusu-antwi) [reply qiskit_superstaq/custom_gates.py:56]: I wasn't sure about this. If you ask IBM to calibrate an `RZXGate` directly, they give the empty schedule for `rads=0`:
image

File: qiskit_superstaq/custom_gates.py (hunk 6 above)
Comment (dowusu-antwi) [reply qiskit_superstaq/custom_gates.py]: It shouldn't be -- done, thanks!

File: qiskit_superstaq/custom_gates.py (hunk 7 above)
Comment (dowusu-antwi) [reply qiskit_superstaq/custom_gates.py]: Done.

File: qiskit_superstaq/custom_gates.py (hunk 8 above)
Comment (dowusu-antwi) [reply qiskit_superstaq/custom_gates.py]: Also shouldn't be -- done. (For some reason I thought to hide the parameter not just for pi/2 (standard AceCR) but also 0.)

File: setup.cfg (hunk 9 above)
Comment (dowusu-antwi) [reply setup.cfg]: Did this accidentally (probably via a merge or `stash pop` or something) -- fixed.

File: qiskit_superstaq/custom_gates.py (hunk 1 above)
Comment (richrines1) [reply qiskit_superstaq/custom_gates.py:52]: i think that makes sense for a rzx (where theta=0 gives you the identity), but afaik acecr(0) shouldn't be an identity? in particular it looks like they're using `rzx(theta, q0) --> x(q0) @ AceCR(theta, q0, q1)`, so for theta=0 the x(q0) cancels the X in between the CRs (and they don't have a sandwiched gate)

Comment (dowusu-antwi) [reply qiskit_superstaq/custom_gates.py]: Done.

File: qiskit_superstaq/custom_gates.py [hunk 10]
```diff
@@ -13,77 +13,74 @@ class AceCR(qiskit.circuit.Gate):
     half-CR ("Z side" and "X side" refer to the two sides of the underlying ZX interactions).
 
     Args:
-        polarity: Should be either "+-" or "-+". Specifies if positive or negative half-CR is first
+        rads: Angle of rotation for CR gate (i.e., twice the angle for each echoed half-CR).
         sandwich_rx_rads: Angle of rotation for an rx gate applied to the "X side" simultaneously
             with the X gate on the "Z side".
         label: An optional label for the constructed Gate
```
Comment (richrines1) [inline qiskit_superstaq/custom_gates.py:19]:
```suggestion
        label: An optional label for the constructed Gate.
```

**Review by richrines1 — APPROVED** [review]
lg w/ one more nit!

**Review by richrines1 — APPROVED** [review]
